cmd/redisoperator/main.go

```diff
@@ -0,0 +1,216 @@
+	flag.StringVar(&c.listenAddr, "listen-address", ":9710", "Address to listen on for metrics.")
+	flag.StringVar(&c.metricsPath, "metrics-path", "/metrics", "Path to serve the metrics.")
+
+	// Parse flags
+	flag.Parse()
+}
+
+func (c *cmdFlags) toRedisOperatorConfig() redisfailover.Config {
+	return redisfailover.Config{
+		Labels:        map[string]string{},
+		ListenAddress: c.listenAddr,
+		MetricsPath:   c.metricsPath,
+	}
+}
+
+// Main is the  main runner.
+type Main struct {
+	flags     *cmdFlags
+	k8sConfig rest.Config
+	logger    log.Logger
+	stopC     chan struct{}
+}
+
+// New returns a Main object.
+func New(logger log.Logger) Main {
+	// Init flags.
+	flgs := &cmdFlags{}
+	flgs.init()
+
+	return Main{
+		logger: logger,
+		flags:  flgs,
+	}
+}
+
+// Run execs the program.
+func (m *Main) Run() error {
+	// Create signal channels.
+	m.stopC = make(chan struct{})
+	errC := make(chan error)
+
+	// Set correct logging.
+	if m.flags.debug {
+		m.logger.Set("debug")
+		m.logger.Debugf("debug mode activated")
+	}
+
+	// Create the metrics client.
+	metricsServer := metrics.NewPrometheusMetrics(m.flags.metricsPath, http.DefaultServeMux)
+
+	// Serve metrics.
+	go func() {
+		log.Infof("Listening on %s for metrics exposure", m.flags.listenAddr)
+		http.ListenAndServe(m.flags.listenAddr, nil)
+	}()
+
+	// Kubernetes clients.
+	stdclient, customclient, aeClientset, err := m.createKubernetesClients()
+	if err != nil {
+		return err
+	}
+
+	// Create kubernetes service.
+	k8sservice := k8s.New(stdclient, customclient, aeClientset, m.logger)
+
+	// Create the redis clients
+	redisClient := redis.New()
+
+	// Create operator and run.
+	redisfailoverOperator := redisfailover.New(m.flags.toRedisOperatorConfig(), k8sservice, redisClient, metricsServer, m.logger)
+	go func() {
+		errC <- redisfailoverOperator.Run(m.stopC)
+	}()
+
+	// Await signals.
+	sigC := m.createSignalCapturer()
+	var finalErr error
+	select {
+	case <-sigC:
+		m.logger.Infof("Signal captured, exiting...")
+	case err := <-errC:
+		m.logger.Errorf("Error received: %s, exiting...", err)
+		finalErr = err
+	}
+
+	m.stop(m.stopC)
+	return finalErr
+}
+
+// loadKubernetesConfig loads kubernetes configuration based on flags.
+func (m *Main) loadKubernetesConfig() (*rest.Config, error) {
+	var cfg *rest.Config
+	// If devel mode then use configuration flag path.
+	if m.flags.development {
+		config, err := clientcmd.BuildConfigFromFlags("", m.flags.kubeConfig)
+		if err != nil {
+			return nil, fmt.Errorf("could not load configuration: %s", err)
+		}
+		cfg = config
+	} else {
+		config, err := rest.InClusterConfig()
+		if err != nil {
+			return nil, fmt.Errorf("error loading kubernetes configuration inside cluster, check app is running outside kubernetes cluster or run in development mode: %s", err)
+		}
+		cfg = config
+	}
+
+	return cfg, nil
+}
+
+func (m *Main) createKubernetesClientWithTimeout(timeout time.Duration) (kubernetes.Interface, error) {
```

Comment:
Is this method (a k8s timeout client) used?

Reply:
I'll delete it on the followings PRs, thanks!